include/lbann/execution_algorithms/batch_functional_inference_algorithm.hpp

```diff
@@ -48,35 +58,110 @@ class batch_functional_inference_algorithm {
   batch_functional_inference_algorithm& operator=(batch_functional_inference_algorithm&& other) = default;
   /** Destructor. */
   virtual ~batch_functional_inference_algorithm() = default;
-  /** Copy training_algorithm. */
-  //  virtual batch_functional_inference_algorithm* copy() const = default;
 
   std::string get_name() const { return "batch_functional_inference"; }
 
+  std::string get_type() const { return "batch_functional_inference"; }
+
   // ===========================================
   // Execution
   // ===========================================
 
-  /** Infer on samples from a data coordinator with a given model. */
-  template <typename TensorDataType>
-  El::AbstractDistMatrix<TensorDataType>
-  infer(model& model,
-        El::AbstractDistMatrix<TensorDataType> const& samples,
-        std::string output_layer,
-        size_t mbs=0);
+  /** @brief Run model inference on samples and return predicted categories.
+   * @param[in] model A trained model
+   * @param[in] samples A distributed matrix containing samples for model input
+   * @param[in] mbs The max mini-batch size
+   * @return Matrix of predicted labels (by index)
+   */
+  template <typename DataT, El::Dist CDist, El::Dist RDist, El::DistWrap DistView, El::Device Device>
+  El::Matrix<int, El::Device::CPU>
+  infer(observer_ptr<model> model,
```

Comment:
I don't think that we need to do this for this PR, but we should move most of this to an impl.hpp file.